### PR TITLE
fix: redirect unauthorized git users to login screen

### PR DESCRIPTION
**Pull request by Kira-Pilot**
https://www.loom.com/share/cfb530114a1e421e8ee388adc83e5f7a?sid=efce074f-034f-43e4-9132-9fc863fb658b

resolves #10902

### CODE REVIEW

File: site/src/pages/LoginPage/SignInForm.tsx [hunk 1]
```diff
@@ -91,7 +91,7 @@ export const SignInForm: FC<React.PropsWithChildren<SignInFormProps>> = ({
         </div>
       )}
 
-      {Boolean(info) && Boolean(error) && (
+      {Boolean(info) && (
```
Comment (Kira-Pilot) [inline site/src/pages/LoginPage/SignInForm.tsx:94]: @aslilac just want to double-check we don't need this error check here.

File: coderd/userauth.go [hunk 2]
```diff
@@ -535,9 +536,7 @@ func (api *API) userOAuth2Github(rw http.ResponseWriter, r *http.Request) {
 			}
 		}
 		if len(selectedMemberships) == 0 {
-			httpapi.Write(ctx, rw, http.StatusUnauthorized, codersdk.Response{
-				Message: "You aren't a member of the authorized Github organizations!",
-			})
+			httpmw.RedirectToLogin(rw, r, &url.URL{Path: "/login"}, "You aren't a member of the authorized Github organizations!")
```
Comment (Kira-Pilot) [inline coderd/userauth.go:539]: if we're ok with this pattern, I can redirect similarly on the other oauth errors.

Comment (deansheather) [reply coderd/userauth.go:539]: One issue with this function is that it sets `?redirect=${currentPath}`, which we don't want to do. If you could change the `httpmw.RedirectToLogin` function to accept a nil `*http.Request` and avoid adding the redirect query param then it should solve the problem

Comment (deansheather) [inline coderd/userauth.go:539]:
```suggestion
			httpmw.RedirectToLogin(rw, r, nil, "You aren't a member of the authorized Github organizations!")
```

dashboardURL defaults to `/login` if nil

File: site/src/pages/UserSettingsPage/SecurityPage/SingleSignOnSection.tsx [hunk 3]
```diff
@@ -62,7 +62,7 @@ export const useSingleSignOnSection = () => {
         // The redirect on success should be back to the login page with a nice message.
         // The user should be logged out if this worked.
         encodeURIComponent(
-          `/login?info=Login type has been changed to ${loginTypeMsg}. Log in again using the new method.`,
+          `/login?message=Login type has been changed to ${loginTypeMsg}. Log in again using the new method.`,
```
Comment (aslilac) [inline site/src/pages/UserSettingsPage/SecurityPage/SingleSignOnSection.tsx:65]: this'll evaluate to "%2Flogin%3Fmessage%3DLogin%20type%20...", which I'm guessing is what we want but this diff doesn't give much context. if it was fine before though it's probably ok!

File: site/src/pages/LoginPage/SignInForm.tsx [hunk 4]
```diff
@@ -91,9 +91,9 @@ export const SignInForm: FC<React.PropsWithChildren<SignInFormProps>> = ({
         </div>
       )}
 
-      {Boolean(info) && Boolean(error) && (
+      {Boolean(message) && (
```
Comment (aslilac) [inline site/src/pages/LoginPage/SignInForm.tsx:94]:
```suggestion
      {message && (
```

nit: technically this cast is superfluous

File: site/src/pages/LoginPage/SignInForm.tsx [hunk 5]
```diff
@@ -63,7 +63,7 @@ export interface SignInFormProps {
   isSigningIn: boolean;
   redirectTo: string;
   error?: unknown;
-  info?: string;
+  message?: string;
```
Comment (aslilac) [inline site/src/pages/LoginPage/SignInForm.tsx:66]: if this was `message?: ReactNode;`...

File: site/src/pages/LoginPage/LoginPageView.tsx [hunk 6]
```diff
@@ -24,7 +24,8 @@ export const LoginPageView: FC<LoginPageViewProps> = ({
   const redirectTo = retrieveRedirect(location.search);
   // This allows messages to be displayed at the top of the sign in form.
   // Helpful for any redirects that want to inform the user of something.
-  const info = new URLSearchParams(location.search).get("info") || undefined;
+  const message =
+    new URLSearchParams(location.search).get("message") || undefined;
```
Comment (aslilac) [inline site/src/pages/LoginPage/LoginPageView.tsx:28]: ...then you wouldn't need this `|| undefined`!

**Review by aslilac — APPROVED** [review]
just some super minor suggestions. looks good!

File: site/src/pages/LoginPage/SignInForm.tsx (hunk 1 above)
Comment (aslilac) [reply site/src/pages/LoginPage/SignInForm.tsx]: yeah, this is probably fine

**Pull request merged**